File: agentic-orchestration-tool/orchestration/ollama_keepalive.py
"""Keep Ollama models resident on the host via periodic /api/generate pings."""

from __future__ import annotations

import os
import sys
import threading
import time
from datetime import datetime, timezone
from typing import Any

import httpx

#: Last keepalive tick — exposed on engine ``/health`` for warm-set diagnostics.
_last_status: dict[str, Any] = {
    "models": [],
    "ok": None,
    "ts": None,
    "base": None,
}


def _env_truthy(name: str, *, default: bool = True) -> bool:
    raw = os.getenv(name)
    if raw is None or not str(raw).strip():
        return default
    return str(raw).strip().lower() not in ("0", "false", "no", "off")


def ollama_keepalive_enabled() -> bool:
    return _env_truthy("AGENTIC_OLLAMA_KEEPALIVE", default=True)


def resolve_ollama_api_base() -> str:
    raw = (
        os.getenv("OLLAMA_API_BASE", "").strip()
        or os.getenv("OLLAMA_HOST", "").strip()
        or "http://127.0.0.1:11434"
    )
    if raw.startswith("http://") or raw.startswith("https://"):
        return raw.rstrip("/")
    return f"http://{raw.rstrip('/')}"


def _normalize_model_tag(raw: str) -> str:
    text = str(raw or "").strip()
    if not text:
        return ""
    lower = text.lower()
    if lower.startswith("ollama/"):
        return text[len("ollama/") :].strip()
    if lower.startswith("ollama:"):
        return text[len("ollama:") :].strip()
    return text


def resolve_keepalive_model_tag() -> str:
    """Single model tag for /api/generate (no ollama/ prefix). Prefer multi-model helper."""
    raw = (
        os.getenv("AGENTIC_OLLAMA_KEEPALIVE_MODEL", "").strip()
        or os.getenv("AGENTIC_PLANNER_MODEL", "").strip()
    )
    return _normalize_model_tag(raw)


def resolve_keepalive_model_tags() -> list[str]:
    """
    Model tags to keep resident.

    ``AGENTIC_OLLAMA_KEEPALIVE_MODELS`` (comma-separated) wins when set — use for a
    meeting pair that shares or lists small chat models. Otherwise falls back to the
    single ``AGENTIC_OLLAMA_KEEPALIVE_MODEL`` / planner model.
    """
    multi = os.getenv("AGENTIC_OLLAMA_KEEPALIVE_MODELS", "").strip()
    if multi:
        out: list[str] = []
        seen: set[str] = set()
        for part in multi.split(","):
            tag = _normalize_model_tag(part)
            key = tag.casefold()
            if tag and key not in seen:
                seen.add(key)
                out.append(tag)
        return out
    single = resolve_keepalive_model_tag()
    return [single] if single else []


def ollama_keepalive_duration() -> str:
    return os.getenv("AGENTIC_OLLAMA_KEEP_ALIVE", "-1").strip() or "-1"


def ollama_keepalive_interval_seconds() -> float:
    raw = os.getenv("AGENTIC_OLLAMA_KEEPALIVE_INTERVAL_MS", "300000").strip()
    try:
        ms = int(raw)
    except ValueError:
        ms = 300_000
    ms = max(30_000, min(ms, 3_600_000))
    return ms / 1000.0


def keepalive_status() -> dict[str, Any]:
    """Snapshot of the last keepalive tick (for ``/health``)."""
    return dict(_last_status)


def _record_status(*, models: list[str], ok: bool | None, base: str) -> None:
    _last_status["models"] = list(models)
    _last_status["ok"] = ok
    _last_status["base"] = base
    _last_status["ts"] = datetime.now(timezone.utc).isoformat()


def ping_ollama_keepalive(*, timeout_seconds: float = 120.0) -> bool:
    if not ollama_keepalive_enabled():
        _record_status(models=[], ok=None, base=resolve_ollama_api_base())
        return False
    models = resolve_keepalive_model_tags()
    base = resolve_ollama_api_base()
    if not models:
        _record_status(models=[], ok=None, base=base)
        return False
    all_ok = True
    try:
        with httpx.Client(timeout=timeout_seconds) as client:
            for model in models:
                body: dict[str, Any] = {
                    "model": model,
                    "prompt": " ",
                    "stream": False,
                    "keep_alive": ollama_keepalive_duration(),
                    "options": {"num_predict": 1},
                }
                res = client.post(f"{base}/api/generate", json=body)
                if not res.is_success:
                    all_ok = False
    except httpx.HTTPError:
        all_ok = False
    _record_status(models=models, ok=all_ok, base=base)
    return all_ok


_timer: threading.Timer | None = None
_in_flight = False
_log_prefix = "(agentic) ollama keep-alive"


def start_ollama_keepalive_loop(*, log_prefix: str | None = None) -> None:
    """Background warmup + interval pings. Safe to call multiple times (no-op after first)."""
    global _timer, _log_prefix
    if _timer is not None:
        return
    if not ollama_keepalive_enabled():
        return
    models = resolve_keepalive_model_tags()
    if not models:
        print(
            f"{_log_prefix}: skipped (set AGENTIC_OLLAMA_KEEPALIVE_MODELS=qwen2.5:3b "
            "or AGENTIC_PLANNER_MODEL=ollama/...)",
            file=sys.stderr,
        )
        return
    if log_prefix:
        _log_prefix = log_prefix

    interval_s = ollama_keepalive_interval_seconds()
    base = resolve_ollama_api_base()
    print(
        f"{_log_prefix}: models={models} base={base} interval_s={interval_s:.0f}",
        file=sys.stderr,
    )

    def tick() -> None:
        global _in_flight, _timer
        if _in_flight:
            _schedule()
            return
        _in_flight = True
        try:
            ok = ping_ollama_keepalive()
            if not ok:
                print(f"{_log_prefix}: ping failed", file=sys.stderr)
        finally:
            _in_flight = False
            _schedule()

    def _schedule() -> None:
        global _timer
        _timer = threading.Timer(interval_s, tick)
        _timer.daemon = True
        _timer.start()

    tick()


def stop_ollama_keepalive_loop() -> None:
    global _timer
    if _timer is not None:
        _timer.cancel()
        _timer = None

File: agentic-orchestration-web/lib/ollama-keepalive.mjs
/**
 * Keep the planner Ollama model loaded via periodic /api/generate pings.
 */

function envTruthy(name, defaultTruthy = true) {
  const raw = process.env[name];
  if (raw == null || String(raw).trim() === "") return defaultTruthy;
  const v = String(raw).trim().toLowerCase();
  return !["0", "false", "no", "off"].includes(v);
}

export function ollamaKeepAliveEnabled() {
  // Resource-sharing broker owns idle unload; forever-keepalive fights that policy.
  const sharing = envTruthy("AGENTIC_OLLAMA_RESOURCE_SHARING", false);
  if (sharing && !envTruthy("AGENTIC_OLLAMA_KEEPALIVE_WITH_SHARING", false)) {
    return false;
  }
  return envTruthy("AGENTIC_OLLAMA_KEEPALIVE", true);
}

export function resolveOllamaApiBase() {
  const raw =
    String(process.env.OLLAMA_API_BASE || "").trim() ||
    String(process.env.OLLAMA_HOST || "").trim() ||
    "http://127.0.0.1:11434";
  if (raw.startsWith("http://") || raw.startsWith("https://")) {
    return raw.replace(/\/+$/, "");
  }
  return `http://${raw.replace(/\/+$/, "")}`;
}

/** Prefer IPv4 loopback — Node often resolves localhost to ::1 first. */
export function normalizeOllamaLoopbackBase(base) {
  return String(base || "").replace(/:\/\/localhost(?=[:/]|$)/gi, "://127.0.0.1");
}

/** Ollama model tag for /api/generate (no ollama/ prefix). */
export function resolvePlannerOllamaModelTag() {
  const raw = String(process.env.AGENTIC_PLANNER_MODEL || "").trim();
  if (!raw) return "";
  const lower = raw.toLowerCase();
  if (lower.startsWith("ollama/")) return raw.slice("ollama/".length).trim();
  if (lower.startsWith("ollama:")) return raw.slice("ollama:".length).trim();
  return "";
}

/** Ollama keep_alive: number seconds, -1 forever, or duration string like "5m". */
export function ollamaKeepAliveDuration() {
  const raw = String(process.env.AGENTIC_OLLAMA_KEEP_ALIVE || "-1").trim();
  const v = raw || "-1";
  // Numeric forever / seconds — must be JSON number (string "-1" → HTTP 400).
  if (/^-?\d+$/.test(v)) return Number(v);
  return v;
}

/** Default 60s — local CPU models unload quickly if idle for minutes. */
export function ollamaKeepAliveIntervalMs() {
  const raw = String(process.env.AGENTIC_OLLAMA_KEEPALIVE_INTERVAL_MS || "60000").trim();
  const n = Number(raw);
  if (!Number.isFinite(n) || n < 15_000) return 60_000;
  return Math.min(n, 3_600_000);
}

export function ollamaKeepAliveMaxAttempts() {
  const raw = String(process.env.AGENTIC_OLLAMA_KEEPALIVE_ATTEMPTS || "3").trim();
  const n = Number(raw);
  if (!Number.isFinite(n) || n < 1) return 3;
  return Math.min(Math.floor(n), 8);
}

async function sleep(ms) {
  await new Promise((r) => setTimeout(r, ms));
}

/** Probe daemon; returns true when /api/tags responds OK. */
export async function pingOllamaTags(opts = {}) {
  const base = normalizeOllamaLoopbackBase(resolveOllamaApiBase());
  const res = await fetch(`${base}/api/tags`, {
    method: "GET",
    signal: opts.signal ?? AbortSignal.timeout(15_000),
  });
  return res.ok;
}

/**
 * Load or refresh model residency. Returns true when Ollama accepted the request.
 * Retries a few times and wakes the daemon via /api/tags first.
 * @param {object} [opts]
 * @param {AbortSignal} [opts.signal]
 */
export async function pingOllamaKeepAlive(opts = {}) {
  if (!ollamaKeepAliveEnabled()) return false;
  const model = resolvePlannerOllamaModelTag();
  if (!model) return false;

  const base = normalizeOllamaLoopbackBase(resolveOllamaApiBase());
  const attempts = ollamaKeepAliveMaxAttempts();
  let lastErr = null;

  for (let i = 0; i < attempts; i++) {
    try {
      // Wake daemon (cheap) before generate — helps after sleep/idle.
      await pingOllamaTags({ signal: opts.signal });

      const url = `${base}/api/generate`;
      const body = {
        model,
        prompt: ".",
        stream: false,
        keep_alive: ollamaKeepAliveDuration(),
        options: { num_predict: 1 },
      };

      const res = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
        signal: opts.signal ?? AbortSignal.timeout(120_000),
      });
      if (res.ok) return true;
      lastErr = new Error(`HTTP ${res.status}`);
    } catch (err) {
      lastErr = err;
    }
    if (i + 1 < attempts) {
      await sleep(500 * (i + 1));
    }
  }
  if (lastErr && opts.throwOnError) throw lastErr;
  return false;
}

let _timer = null;
let _inFlight = false;
/** When >0, skip keep-alive pings so chat/orchestrate owns the Ollama slot. */
let _orchestrateBusy = 0;

export function beginOrchestrateOllamaBusy() {
  _orchestrateBusy += 1;
}

export function endOrchestrateOllamaBusy() {
  _orchestrateBusy = Math.max(0, _orchestrateBusy - 1);
}

/** Start warmup + interval keep-alive (no-op when disabled or planner is not Ollama). */
export function startOllamaKeepAliveLoop() {
  if (_timer) return;
  if (!ollamaKeepAliveEnabled()) return;
  const model = resolvePlannerOllamaModelTag();
  if (!model) {
    console.error(
      "[agentic-orchestration-web] Ollama keep-alive skipped (AGENTIC_PLANNER_MODEL is not ollama/...)",
    );
    return;
  }

  const intervalMs = ollamaKeepAliveIntervalMs();
  const base = normalizeOllamaLoopbackBase(resolveOllamaApiBase());
  console.error(
    `[agentic-orchestration-web] Ollama keep-alive: model=${model} base=${base} interval_ms=${intervalMs}`,
  );

  const tick = async () => {
    if (_inFlight || _orchestrateBusy > 0) return;
    _inFlight = true;
    try {
      const ok = await pingOllamaKeepAlive();
      if (!ok) {
        console.error("[agentic-orchestration-web] Ollama keep-alive ping failed");
      }
    } catch (err) {
      console.error(
        `[agentic-orchestration-web] Ollama keep-alive error: ${err instanceof Error ? err.message : err}`,
      );
    } finally {
      _inFlight = false;
    }
  };

  tick();
  _timer = setInterval(tick, intervalMs);
  if (typeof _timer.unref === "function") _timer.unref();
}

export function stopOllamaKeepAliveLoop() {
  if (_timer) {
    clearInterval(_timer);
    _timer = null;
  }
}
